Getting Started with JavaScript

mkaz opened the pull request:
## Description

An introductory tutorial for getting started working with JavaScript and the Block Editor. A lot of the documentation is just snippets of JavaScript code assuming the developers know what to do with it. This tutorial aims to get developers who may not be familiar started off.

See #11392

I welcome any additional feedback, expanding, and building on this as a start.

Also, I am not sure how the documentation manages the multiple pages and navigation so this probably could be broken up into other pages like the block tutorial.

Online Preview: https://github.com/WordPress/gutenberg/tree/docs/11392/javascript/docs/designers-developers/developers/tutorials/javascript

mkaz committed:
Getting Started with JavaScript

joyously commented:
Please use standard practices, such as a consistent prefix, and one that is not just two letters (my).
People get confused with where prefixing is needed, and why the prefix can't have a dash.

Also, please be consistent in the PHP with regard to order. Either function and then add_action, or add_action and then function, but use the same way every time. It keeps it easier to digest. I personally prefer to define the function and then add_action, but that's just how I think.

You might ought to mention the version number parameter for enqueuing scripts, but especially for styles.

This tutorial is for a script in the editor, but your style is loaded only on the front end. That will be confusing, unless your next section is about the different actions available for loading styles.

mkaz committed:
Reorder add_action/function

mkaz committed:
Update prefix to use myguten, instead of just my

mkaz commented:
@joyously Thanks for the feedback. I updated the prefix and ordering of the action/functions.

> You might ought to mention the version number parameter for enqueuing scripts, but especially for styles.

I was debating how much WordPress parts to go into, I ink out to the relevant documentation for each section. It's always a hard balance to keep things brief to not overwhelm people, but also complete.

> This tutorial is for a script in the editor, but your style is loaded only on the front end. That will be confusing, unless your next section is about the different actions available for loading styles.

The block style modification adds a class to blockquotes, which isn't really visisble unless you look at the source. So I thought it would be nice to be able to see it doing something, which is why I added the CSS part, not sure if that is overly confusing.

As for next part, not sure how deep to go with it, the idea was to get setup and show an example of using the documentation to apply the change. I'm all for adding to it to help users further.

joyously commented:
>  It's always a hard balance to keep things brief to not overwhelm people

I agree, however they will be confounded when they update and don't see the change because it's cached. I've seen so many theme authors leave version numbers off and then have to field support questions about clearing cache etc. so that their updates take effect.

>  see it doing something, which is why I added the CSS part

It's not too confusing except that everyone always says they want the back to match the front, so why wouldn't you load the CSS in both?

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
```diff
@@ -0,0 +1,108 @@
+
+# Getting Started with JavaScript
+
+The purpose of this tutorial is to step through getting started with JavaScript and WordPress. The Gutenberg documentation contains information on the APIs available, but as a reference and not necessarily full working examples. The goal of this tutorial is get you comfortable on how to use these snippets of code.
```
Would it be worthwhile to have a short explanation of "What is JavaScript"?

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
````diff
@@ -0,0 +1,108 @@
+
+```js
+console.log( "I'm loaded!" );
+```
````
Probably worth recapping each of the files you have at this point.

You could also offer a direct download of this basic stub plugin to help people get going.

chrisvanpatten replied:
I am wondering if we should use the `gutenberg-examples` repo for storing code associated with specific tutorials like this?

mkaz replied:
The tutorial was updated and a recap added.

I like the idea of a Github repo to put the code, but the `gutenberg-examples` is so tailored for it being a plugin, it might not be the right fit. A `gutenberg-tutorials` ?  We can discuss further in #docs maybe even move content out of handbook and link to make them more stand-alone.

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
```diff
@@ -0,0 +1,108 @@
+function myguten_enqueue( $hook ) {
+	wp_enqueue_script( 'myguten-script',
+		plugins_url( 'myguten.js', __FILE__ ),
+		array( 'wp-blocks', 'wp-element' )
```
Is it worth having a code comment explaining what each of these function arguments is? I think the note you have below is less obvious.

mkaz replied:
Rearranged and updated information on dependencies.

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
```diff
@@ -0,0 +1,108 @@
+
+Now go to any page within `/wp-admin/` and check your browser Developer Tools, and you should see the message in your console log. See [Mozilla's What are browser developer tools?](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools) if you need more information, the area to become most familiar with is [The JavaScript console](https://developer.mozill ...
+
+The script enqueuing used the `admin_enqueue_scripts` hook which only loads the JavaScript within the wp-admin section. So if you navigate to any front-end page or article on your site, you will not see the message.
```
I don't think `admin_enqueue_scripts` is the correct action for enqueuing block assets...

mkaz replied:
You're right, updated. I was reading through the core plugin/theme documentation and trying to use those examples the user might be familiar with to match the same ideas to working with Gutenberg.

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
```diff
@@ -0,0 +1,108 @@
+
+## JavaScript versions and building
+
+The Gutenberg Handbook shows examples in two syntaxes ES5 and ESNext, note ESNext is sometimes referred to as ES6. The ES5 code is compatible with almost all current browsers.
```
```suggestion
The Gutenberg Handbook shows examples in two variations of JavaScript, ES5 and ESNext. What's even more confusing is that ESNext is sometimes referred to as ES6. The ES5 code is compatible with almost all current browsers.
```

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
```diff
@@ -0,0 +1,108 @@
+
+The ESNext syntax is compatible with most modern browsers, but unfortunately not IE11. Additionally, the ESNext code examples include JSX syntax, which requires a build step using webpack to transform into code that can be run in your browser.
+
+For this tutorial, all examples will be in straight ES5 code that can be run straight in your browser and does not require an additional build step.
```
```suggestion
For this tutorial, all examples are written as the ES5 variation of JavaScript that can be run straight in your browser and does not require an additional build step.
```

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
```diff
@@ -0,0 +1,108 @@
+
+Let's look at using the [Block Style Variation example](../../../../../docs/designers-developers/developers/filters/block-filters/#block-style-variations).
+
+Update the file `myguten.js` with:
```
```suggestion
Replace the existing `console.log()` code in your `myguten.js` with:
```

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
````diff
@@ -0,0 +1,108 @@
+} );
+```
+
+You can remove the previous `console.log` if you like.
````
```suggestion

```

mkaz replied:
Ah yes, "Replace". Thanks, I found the phrasing awkward when I wrote it but a better way didn't come to mind

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
```diff
@@ -0,0 +1,108 @@
+
+You can remove the previous `console.log` if you like.
+
+After you add the code, save the file, and start a new post in the Block Editor.
```
```suggestion
After you add the `wp.blocks.registerBlockStyle` code, save the `myguten.js` file, and then create a new post in the Block Editor.
```

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
```diff
@@ -0,0 +1,108 @@
+
+After you add the code, save the file, and start a new post in the Block Editor.
+
+Add a quote block, and in the right sidebar under Styles, you should see your new Fancy Quote style listed. You can go back to the JavaScript and change the label to "Fancy Pants" and reload, and you should see the new label.
```
Worth including a screenshot here?

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
```diff
@@ -0,0 +1,108 @@
+function myguten_stylesheet() {
+	wp_enqueue_style( 'myguten-style', plugins_url( 'style.css', __FILE__ ) );
+}
+add_action( 'wp_enqueue_scripts', 'myguten_stylesheet' );
```
```suggestion
add_action( 'enqueue_block_assets', 'myguten_stylesheet' );
```

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
```diff
@@ -0,0 +1,108 @@
+
+Previewing or Publishing the post, you will not see a visible change. However, if you look at the source, you should see `is-style-fancy-quote` class name attached to your quote.
+
+You could create a `style.css` file with:
```
```suggestion
Go ahead and create a `style.css` file with:
```

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
````diff
@@ -0,0 +1,108 @@
+and enqueue the CSS by adding the following to your `myguten-plugin.php`:
+
+```php
+function myguten_stylesheet() {
````
Can this function have PHPDoc too, explaining what it's doing?

mkaz replied:
I've broken up the code and changed things around with more explanations above the code. My concern is if it were as a comment in code people will miss it. I've also included links to all referred to functions for any additional details are available.

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
````diff
@@ -0,0 +1,108 @@
+add_action( 'wp_enqueue_scripts', 'myguten_stylesheet' );
+```
+
+And then when you view the page, you should see it in a very large font.
````
```suggestion
And then when you view the page, you will see it in a very large font.
```

danielbachhuber commented on docs/designers-developers/developers/tutorials/javascript/readme.md:
```diff
@@ -0,0 +1,108 @@
+
+And then when you view the page, you should see it in a very large font.
+
+
```
Would it be worthwhile to include some "But it's not working?" debugging steps?

mkaz replied:
👍

danielbachhuber reviewed (commented):
Great start! I've left some comments and suggestions.

If you have the bandwidth, it think it'd be really worthwhile to "user test" this documentation with a layperson at WCUS. Specifically, sit them down at a laptop, have them follow the documentation, and observe their usage to see what sort of questions they run into.

danielbachhuber committed:
Update docs/designers-developers/developers/tutorials/javascript/readme.md

Co-Authored-By: mkaz <marcus@mkaz.com>

mkaz committed:
Update to block_editor_assets

mkaz committed:
Updates per danielbachhuber review.

mkaz commented:
@danielbachhuber Thanks so much for all the feedback, I really appreciate it! I updated with your change suggestions. I'll take another pass through for the code docs and additional ideas.

> If you have the bandwidth, it think it'd be really worthwhile to "user test" this documentation with a layperson at WCUS. 

Good suggestion, but I'm unfortunately not at WCUS. It would be great to see someone new walk through and see what additional questions they might have. It is always tricky trying to guess where someone might get stuck.

mkaz committed:
Updates, and broken up into sections

mkaz committed:
Add Troubleshooting

nosolosw commented on docs/designers-developers/developers/tutorials/javascript/plugins-background.md:
````diff
@@ -0,0 +1,17 @@
+/*
+Plugin Name: My Guten Plugin
+*/
+```
````
Should this comment use [DocBlock](https://en.wikipedia.org/wiki/PHPDoc#Components_of_PHPDoc) instead? Like

```
/**
 * Plugin Name: My Guten Plugin
 */

```
My eyes seem to be more used to this kind of comments when it comes to WordPress plugins, but I'm unsure if that's a recommendation written anywhere.

TimothyBJacobs replied:
Either works, but the standard is `/*` for the header: https://developer.wordpress.org/plugins/the-basics/header-requirements/

mkaz replied:
I'll leave as standard, thanks @TimothyBJacobs

nosolosw commented on docs/designers-developers/developers/tutorials/javascript/loading-javascript.md:
```diff
@@ -0,0 +1,48 @@
+function myguten_enqueue() {
+	wp_enqueue_script( 'myguten-script',
+		plugins_url( 'myguten.js', __FILE__ ),
+		array( 'wp-blocks', 'wp-element' )
```
I dig this step-by-step approach! It's very clear.

For the sake of creating a _Minimum Viable Plugin_, how about removing these dependencies. They're not required just yet to write to the console.

nosolosw commented on docs/designers-developers/developers/tutorials/javascript/loading-javascript.md:
```diff
@@ -0,0 +1,48 @@
+
+The script enqueuing uses the `enqueue_block_editor_assets` hook which only loads the JavaScript when the block editor loads. So if you navigate to any front-end page or article on your site, you will not see the message.
+
+The last argument in the `wp_enqueue_script()` function is an array of dependencies. All of the Gutenberg packages are registered and can be loaded by specifying them in the array, the blocks and elements packages are shown above, as they are two common ones used.
```
We could remove this part about dependencies and add it back in the first example that extends the editor.

mkaz replied:
I rearranged the enqueue with two steps explaining it more clearly in the extension part.

nosolosw commented on docs/designers-developers/developers/tutorials/javascript/loading-javascript.md:
```diff
@@ -0,0 +1,48 @@
+The last argument in the `wp_enqueue_script()` function is an array of dependencies. All of the Gutenberg packages are registered and can be loaded by specifying them in the array, the blocks and elements packages are shown above, as they are two common ones used.
+
+
+**Note for Theme Developers:**  The above method of enqueing is used for plugins, if you are extending Gutenberg for your theme there is a minor difference, you will use the `get_template_directory_uri()` function instead of `plugins_url()`. So for a theme, the enqueue example is:
```
:heart:

nosolosw commented on docs/designers-developers/developers/tutorials/javascript/loading-javascript.md:
```diff
@@ -0,0 +1,48 @@
+function myguten_enqueue() {
+	wp_enqueue_script( 'myguten-script',
+		get_template_directory_uri() . '/myguten.js',
+		array( 'wp-blocks', 'wp-element' )
```
Same comment here about dependencies.

nosolosw commented on docs/designers-developers/developers/tutorials/javascript/extending-the-block-editor.md:
````diff
@@ -0,0 +1,41 @@
+Replace the existing `console.log()` code in your `myguten.js` file with:
+
+```js
+wp.blocks.registerBlockStyle( 'core/quote', {
````
After this piece of code, we could introduce two things: 

- that WordPress makes Gutenberg packages available in the browser through the `wp.` namespace.
- that our plugin needs to tell WordPress what packages we want to use (add dependencies to the enqueueing mechanism).

GitHub still don't let me suggest anything longer than oneliners, but here's a suggestion:

---

Replace the existing `console.log()` code in your `myguten.js` file with:

```js
wp.blocks.registerBlockStyle( 'core/quote', {
    name: 'fancy-quote',
    label: 'Fancy Quote',
} );
```

Update the code in `guten-plugin.php` to be:

```php
/**
 * Plugin Name: My Guten Plugin
 */
function myguten_enqueue() {
	wp_enqueue_script( 'myguten-script',
		plugins_url( 'myguten.js', __FILE__ ),
		array( 'wp-blocks' )
	);
}
add_action( 'enqueue_block_editor_assets', 'myguten_enqueue' );
```

In the [`wp_enqueue_script`](https://developer.wordpress.org/reference/functions/wp_enqueue_script/) function, we declare that our plugin depends on the wp-blocks package (see [Packages](https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/)). WordPress will make that available for us using the `wp` namespace, so, in the browser we can use `wp.blocks` to access anything that package exports.

After these changes, save the files, and then create a new post in the Block Editor.

mkaz replied:
Thanks, this is great, I reworded it some and updated to fit with the flow.

nosolosw commented on docs/designers-developers/developers/tutorials/javascript/troubleshooting.md:
```diff
@@ -0,0 +1,18 @@
+
+If you are not seeing your changes, another place to check is to confirm your JavaScript file is being enqueued. You can look at the page source, right-click View Page Source, and look for the `<script>` tag that loads your file. In our example, you would searching for `myguten.js` and confirm it is being loaded.
+
+If you do not see the file being loaded, doublecheck the enqueue function is correct. You can also check your server logs to see if there is an issue.
```
Perhaps adding a section about dependencies. Arguably it has been more of a trouble for developers that relied on older versions of Gutenberg that didn't require the plugins to specify them so the change took them by surprise. Can still be useful for new devs, though.

Suggestion: 

---

### Confirm all dependencies are loaded

The console log should show an error if a dependency your JavaScript code uses has not been declared and loaded in the browser. In our example, if `myguten.js` script is enqueued without declaring its dependencies such as:

	wp_enqueue_script( 'myguten-script',
		plugins_url( 'myguten.js', __FILE__ )
	);

the console log will include this error:

![screenshot from 2018-12-11 19-36-00](https://user-images.githubusercontent.com/583546/49822139-8a212480-fd7c-11e8-8d77-ff1e36ba54c6.png)

mkaz replied:
Added. Thanks!

nosolosw commented on docs/designers-developers/developers/tutorials/javascript/versions-and-building.md:
```diff
@@ -0,0 +1,14 @@
+
+The Gutenberg Handbook shows JavaScript examples in two syntaxes ES5 and ESNext. These are version names for the JavaScript language standard definitions. You may also see elsewhere the versions ES6, or ECMAScript 2015 mentioned. Ses the [ECMAScript](https://en.wikipedia.org/wiki/ECMAScript) Wikipedia article for all the details.
+
+ES5 code is compatible with WordPress's minimum [target for browser compatibility](https://make.wordpress.org/core/2017/04/23/target-browser-coverage/).
```
We may want to point to the [browser support handbook](https://make.wordpress.org/core/handbook/best-practices/browser-support/) or [the Gutenberg FAQ](https://wordpress.org/gutenberg/handbook/designers-developers/faq/#what-browsers-will-gutenberg-support) instead, they seem more permanent. I'd say the core handbook is probably best.

mkaz replied:
👍Updated the link.

nosolosw commented on docs/designers-developers/developers/tutorials/javascript/versions-and-building.md:
```diff
@@ -0,0 +1,14 @@
+
+## JavaScript versions and build step
```
This section is important but feels disconnected and left me wondered _how one should go about the build step?_. I know Gutenberg is agnostic about what tools third-party authors use, but I still feel this doc should direct them to the best resources available at the moment. It can be changed later if/when Gutenberg itself provides this. I'd include here a reference to the [Create Gutenberg Block](https://github.com/ahmadawais/create-guten-block) project.

mkaz replied:
yeh, this was tricky. I wanted to explain what they might see in the documentation but try to keep it simple.  It is probably in the wrong spot in my table of contents, maybe put it after troubelshooting and expand.

I'm a big fan of getting the person to quick wins as they go through a tutorial.
Do this. Run. See something.
Now add this. Run. See new something.
-> Repeat.

So including the extra info in the middle confuses it. 
I think I'll reorganize it to the end and can expand.

nosolosw commented on docs/designers-developers/developers/tutorials/javascript/versions-and-building.md:
```diff
@@ -0,0 +1,14 @@
+
+The ESNext version is dynamic and refers to the next language defintions, which will always be ahead of browser support. An extra build step is required to transform the code to a syntax that works in all browsers. Webpack is the tool that performs this transformation step.
+
+Additionally, the ESNext code examples in the Gutenberg handbook include JSX syntax, which is easier to read and write, but likewise requires the build step using webpack and babel to transform into compatible code.
```
```suggestion
Additionally, the ESNext code examples in the Gutenberg handbook include [JSX syntax](https://reactjs.org/docs/introducing-jsx.html), which is easier to read and write, but likewise requires the build step using webpack and babel to transform into compatible code.
```

nosolosw commented on docs/designers-developers/developers/tutorials/javascript/versions-and-building.md:
```diff
@@ -0,0 +1,14 @@
+
+For this tutorial, all examples are written compatible to ES5 definition of JavaScript. This code can be run straight in your browser and does not require an additional build step.
+
+The ESNext version is dynamic and refers to the next language defintions, which will always be ahead of browser support. An extra build step is required to transform the code to a syntax that works in all browsers. Webpack is the tool that performs this transformation step.
```
```suggestion
The ESNext version is dynamic and refers to the next language definitions, which will always be ahead of browser support. An extra build step is required to transform the code to a syntax that works in all browsers. Webpack and babel are the tools that perform this transformation step.
```

mkaz committed:
Updates with suggestion. Props to @nosolosw

nosolosw committed:
Update docs/designers-developers/developers/tutorials/javascript/versions-and-building.md

Co-Authored-By: mkaz <marcus@mkaz.com>

nosolosw committed:
Update docs/designers-developers/developers/tutorials/javascript/versions-and-building.md

Co-Authored-By: mkaz <marcus@mkaz.com>

mkaz committed:
Add screenshot with style

mkaz committed:
Reorder TOC, move versions to end

mkaz committed:
Update link for browser support

mkaz committed:
Add check for dependency to troubleshooting section

mkaz commented on docs/designers-developers/developers/tutorials/javascript/extending-the-block-editor.md:
````diff
@@ -0,0 +1,41 @@
+
+```css
+.is-style-fancy-quote {
+	font-size: 64px;
````
This was a great suggestion, switched it to use color and works great in Editor and Published. This also addresses one of @joyously comments I wanted to circle back to. I've included screenshots too.

mkaz committed:
General edits for spelling and flow

oandregal reviewed (approved):
This is great work, bravo!

nosolosw commented:
Merging still seems to be blocked, so unsure how to go about this. cc @chrisvanpatten I'm tagging this as 4.8 in case we have to wait for landing this PR.

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/extending-the-block-editor.md:
```diff
@@ -0,0 +1,67 @@
+
+## Extending the Block Editor
```
This should be an h1, and the opening blank newline should be removed.

mkaz replied:
Updated here and for each sub page in 6672b8340

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/versions-and-building.md:
```diff
@@ -0,0 +1,13 @@
+
+## JavaScript versions and build step
```
this should be an h1 (`#`) with no newline at the top

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/versions-and-building.md:
```diff
@@ -0,0 +1,13 @@
+
+## JavaScript versions and build step
+
+The Gutenberg Handbook shows JavaScript examples in two syntaxes ES5 and ESNext. These are version names for the JavaScript language standard definitions. You may also see elsewhere the versions ES6, or ECMAScript 2015 mentioned. Ses the [ECMAScript](https://en.wikipedia.org/wiki/ECMAScript) Wikipedia article for all the details.
```
```suggestion
The Gutenberg Handbook shows JavaScript examples in two syntaxes: ES5 and ESNext. These are version names for the JavaScript language standard definitions. You may also see elsewhere the names ES6 or ECMAScript 2015 mentioned. See the [ECMAScript](https://en.wikipedia.org/wiki/ECMAScript) Wikipedia article for all the details.
```

mkaz committed:
Update Table of Contents and generated manifest

mkaz commented:
@gziolo The pages will be auto linked with prev/next links at the bottom when published in the handbook. 

You can see the [Blocks Tutorial example here](https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/block-tutorial/)

I updated the table of contents and generated the manifest which makes sure that happens.

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/versions-and-building.md:
```diff
@@ -0,0 +1,13 @@
+
+The ESNext version is dynamic and refers to the next language definitions, which will always be ahead of browser support. An extra build step is required to transform the code to a syntax that works in all browsers. Webpack and babel are the tools that perform this transformation step.
+
+Additionally, the ESNext code examples in the Gutenberg handbook include [JSX syntax](https://reactjs.org/docs/introducing-jsx.html), which is easier to read and write, but likewise requires the build step using webpack and babel to transform into compatible code.
```
These paragraphs should be reorganised a bit. Here's something to start from:

> ES5 code is compatible with WordPress's minimum [target for browser support](https://make.wordpress.org/core/handbook/best-practices/browser-support/).
>
> "ESNext" doesn't refer to a specific version of JavaScript, but is "dynamic" and refers to the next language definitions, whatever they might be. Because some browsers won't support these features yet (because they're new or proposed), an extra build step is required to transform the code to a syntax that works in all browsers. Webpack and babel are the tools that perform this transformation step.
>
> Additionally, the ESNext code examples in the Gutenberg handbook include [JSX syntax](https://reactjs.org/docs/introducing-jsx.html), a syntax that blends HTML and JavaScript. It makes it easier to read and write markup to be displayed in the browser, but likewise requires the build step using webpack and babel to transform into compatible code.
> 
> For simplicity, this tutorial uses the ES5 definition of JavaScript, without JSX. This code can be run straight in your browser and does not require an additional build step.

mkaz replied:
👍

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/troubleshooting.md:
```diff
@@ -0,0 +1,35 @@
+
+### Confirm JavaScript is Loading
+
+If you are not seeing your changes, another place to check is to confirm your JavaScript file is being enqueued. You can look at the page source, right-click View Page Source, and look for the `<script>` tag that loads your file. In our example, you would searching for `myguten.js` and confirm it is being loaded.
```
```suggestion
If you are not seeing your changes, check that your JavaScript file is being enqueued. Open the page source in your browser's web inspector (some browsers may allow you to view the page source by right clicking on the page and selecting "View Page Source"), and look for the `<script>` tag that loads your file. In our example, you would search for `myguten.js` and confirm it is being loaded.
```

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/troubleshooting.md:
```diff
@@ -0,0 +1,35 @@
+
+The console log will show an error if a dependency your JavaScript code uses has not been declared and loaded in the browser. In the example, if `myguten.js` script is enqueued without declaring the `wp-block` dependency, the console log will show:
+
+<img src="../../../../../docs/designers-developers/developers/tutorials/javascript/error-blocks-undefined.png" width=448 title="error wp.blocks is undefined"/>
```
Can we use the markdown image format here?

mkaz replied:
I tried but the image comes in too wide since it was created on a @2x screen.

Also tried using markdown formatting here: https://gist.github.com/uupaa/f77d2bcf4dc7a294d109 but they didn't work either. 

Open to suggestions, I can size it down but then won't look good on high dpi screens.

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/troubleshooting.md:
```diff
@@ -0,0 +1,35 @@
+
+<img src="../../../../../docs/designers-developers/developers/tutorials/javascript/error-blocks-undefined.png" width=448 title="error wp.blocks is undefined"/>
+
+
```
Extra newline

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/troubleshooting.md:
```diff
@@ -0,0 +1,35 @@
+## Troubleshooting
```
Should be `#`

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/troubleshooting.md:
```diff
@@ -0,0 +1,35 @@
+## Troubleshooting
+
+It's not working, what do I do?
```
```suggestion
If you're having trouble getting your code to work, here are a few ways to troubleshoot.
```

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/troubleshooting.md:
```diff
@@ -0,0 +1,35 @@
+
+### Console Log
+
+The console log is a JavaScript developer's best friend, it is a good practice to work with it open. If there are any errors this is where they will show up. See Mozilla's [JavaScript console](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools#The_JavaScript_console) documentation for more.
```
```suggestion
The console log is a JavaScript developer's best friend. It is a good practice to work with it open, as it collects errors and notices into one place. See Mozilla's [JavaScript console](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools#The_JavaScript_console) documentation for more.
```

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/troubleshooting.md:
```diff
@@ -0,0 +1,35 @@
+
+The console log is a JavaScript developer's best friend, it is a good practice to work with it open. If there are any errors this is where they will show up. See Mozilla's [JavaScript console](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools#The_JavaScript_console) documentation for more.
+
+Check the JavaScript console for any errors. Here is an example, which shows a syntax error on line 6.
```
```suggestion
Your first step in debugging should be to check the JavaScript console for any errors. Here is an example, which shows a syntax error on line 6.
```

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/troubleshooting.md:
```diff
@@ -0,0 +1,35 @@
+![console error](../../../../../docs/designers-developers/developers/tutorials/javascript/console-error.png)
+
+
+### Confirm all dependencies are loaded
```
Can you swap this question below the next one?

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/troubleshooting.md:
```diff
@@ -0,0 +1,35 @@
+
+### Confirm all dependencies are loaded
+
+The console log will show an error if a dependency your JavaScript code uses has not been declared and loaded in the browser. In the example, if `myguten.js` script is enqueued without declaring the `wp-block` dependency, the console log will show:
```
Is it `wp-block` or `wp-blocks`?

mkaz replied:
good catch!

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/troubleshooting.md:
````diff
@@ -0,0 +1,35 @@
+);
+```
+
+
````
extra newline not needed

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/troubleshooting.md:
````diff
@@ -0,0 +1,35 @@
+```
+
+
+### Confirm JavaScript is Loading
````
all these questions should be h2, not h3

chrisvanpatten reviewed (changes requested):
I have a partial review here. More to review, but I have other things to work on this AM and wanted to get you as much as I could quicker :)

mkaz committed:
Fix up header tags consistency and spaces

mkaz committed:
Edits per @chrisvanpatten review

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/extending-the-block-editor.md:
````diff
@@ -0,0 +1,66 @@
+add_action( 'enqueue_block_editor_assets', 'myguten_enqueue' );
+```
+
+The last argument in the `wp_enqueue_script()` function is an array of dependencies. WordPress makes available packages under the wp namespace. In the example, you use `wp.blocks` to access the items the blocks package exports, in this case the `registerBlockStyle()` function.
````
```suggestion
The last argument in the `wp_enqueue_script()` function is an array of dependencies. WordPress makes packages available under the `wp` namespace. In the example, you use `wp.blocks` to access the items that the blocks package exports (in this case the `registerBlockStyle()` function).
```

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/loading-javascript.md:
````diff
@@ -0,0 +1,42 @@
+
+```php
+function myguten_enqueue() {
+	wp_enqueue_script( 'myguten-script',
````
I think per core code style this should be

```
wp_enqueue_script(
    'my-guten-script',
```

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/loading-javascript.md:
````diff
@@ -0,0 +1,42 @@
+console.log( "I'm loaded!" );
+```
+
+Now start a new post in the block editor, and check your browser Developer Tools. You should see the message in your console log. See [Mozilla's What are browser developer tools?](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools) if you need more information, the area to become most familiar with is [The JavaScript console](https://developer.mozilla ...
````
```suggestion
Next, create a new post in the block editor.

We'll check the JavaScript console in your browser's Developer Tools, to see if the message is displayed. If you're not sure what developer tools are, Mozilla's ["What are browser developer tools?"](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools) documentation provides more information, including more background on the [JavaScript console](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools#The_JavaScript_console).

If your code is registered and enqueued correctly, you should see a message in your console:
```

_Then it would be nice to show a screenshot here of the successful console message._

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/loading-javascript.md:
```diff
@@ -0,0 +1,42 @@
+Now start a new post in the block editor, and check your browser Developer Tools. You should see the message in your console log. See [Mozilla's What are browser developer tools?](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools) if you need more information, the area to become most familiar with is [The JavaScript console](https://developer.mozilla ...
+
+
+**Note for Theme Developers:**  The above method of enqueing is used for plugins, if you are extending the Block Editor for your theme there is a minor difference, you will use the `get_template_directory_uri()` function instead of `plugins_url()`. So for a theme, the enqueue example is:
```
```suggestion
**Note for Theme Developers:**  The above method of enqueing is used for plugins. If you are extending the Block Editor for your theme there is a minor difference, you will use the `get_template_directory_uri()` function instead of `plugins_url()`. So for a theme, the enqueue example is:
```

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/loading-javascript.md:
````diff
@@ -0,0 +1,42 @@
+
+```php
+function myguten_enqueue() {
+	wp_enqueue_script( 'myguten-script',
````
same code formatting note as above

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/loading-javascript.md:
```diff
@@ -0,0 +1,42 @@
+
+### Recap
+
+At this point, you have a plugin in the directory `wp-content/plugins/myguten-plugin` with two files. The PHP server-side code in `myguten-plugin.php`, and the JavaScript which runs in the browser in `myguten.js`.
```
```suggestion
At this point, you have a plugin in the directory `wp-content/plugins/myguten-plugin` with two files: the PHP server-side code in `myguten-plugin.php`, and the JavaScript which runs in the browser in `myguten.js`.
```

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/plugins-background.md:
```diff
@@ -0,0 +1,16 @@
+# Plugins Background
+
+The primary means of extending WordPress remains the plugin, see [Plugin Basics](https://developer.wordpress.org/plugins/the-basics/) for more details on developing plugins for WordPress. The quick start is to create a new directory in `wp-content/plugins/` to hold your plugin code, for this example you can call it `myguten-plugin`.
```
```suggestion
The primary means of extending WordPress is the plugin. WordPress' [Plugin Basics](https://developer.wordpress.org/plugins/the-basics/) documentation provides for more details on building a plugin. The quickest way to start is to create a new directory in `wp-content/plugins/` to hold your plugin code, for this example you can call it `myguten-plugin`.
```

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/plugins-background.md:
````diff
@@ -0,0 +1,16 @@
+*/
+```
+
+So you should have a directory `wp-content/plugins/myguten-plugin/` which has the single file `myguten-plugin.php`. Once that is in place, go to your plugins list in `wp-admin` and you should see your plugin listed.
````
```suggestion
To summarize, you should have a directory `wp-content/plugins/myguten-plugin/` which has the single file `myguten-plugin.php`. Once that is in place, go to your plugins list in `wp-admin` and you should see your plugin listed.
```

chrisvanpatten reviewed (changes requested):
Some language tweaks! Looking good and I think it'll be an easy approval after these are resolved.

chrisvanpatten committed:
Update docs/designers-developers/developers/tutorials/javascript/extending-the-block-editor.md

Co-Authored-By: mkaz <marcus@mkaz.com>

chrisvanpatten committed:
Update docs/designers-developers/developers/tutorials/javascript/loading-javascript.md

Co-Authored-By: mkaz <marcus@mkaz.com>

chrisvanpatten committed:
Update docs/designers-developers/developers/tutorials/javascript/loading-javascript.md

Co-Authored-By: mkaz <marcus@mkaz.com>

chrisvanpatten committed:
Update docs/designers-developers/developers/tutorials/javascript/loading-javascript.md

Co-Authored-By: mkaz <marcus@mkaz.com>

chrisvanpatten committed:
Update docs/designers-developers/developers/tutorials/javascript/plugins-background.md

Co-Authored-By: mkaz <marcus@mkaz.com>

chrisvanpatten committed:
Update docs/designers-developers/developers/tutorials/javascript/plugins-background.md

Co-Authored-By: mkaz <marcus@mkaz.com>

mkaz committed:
Move screenshots to assets directory, add additional screenshot for success

mkaz committed:
Move block style variant to assets

mkaz committed:
Formating

chrisvanpatten commented on docs/designers-developers/developers/tutorials/javascript/loading-javascript.md:
```diff
@@ -27,12 +28,14 @@ We'll check the JavaScript console in your browser's Developer Tools, to see if
 
 If your code is registered and enqueued correctly, you should see a message in your console:
 
+![Console Log Message Success](../../../../../docs/designers-developers/assets/js-tutorial-console-log-success.png)
```
With image assets I think we're cool to make them full URLs (and in fact I believe we have to right now; the handbook won't sync images and I think needs the raw URLs linked here)

mkaz replied:
so the URL from Github, like so?

`https://raw.githubusercontent.com/WordPress/gutenberg/docs/11392/javascript/docs/designers-developers/assets/js-tutorial-console-log-error.png`

The above includes the branch, so changes after commit

chrisvanpatten replied:
Yep, you can see how we do it here: https://raw.githubusercontent.com/WordPress/gutenberg/master/docs/designers-developers/designers/block-design.md

mkaz replied:
Updated. 

The images won't show in PR view, since they aren't in master until after merge.

chrisvanpatten reviewed (commented):
SO close. Images need to be full absolute URLs, not just paths. Otherwise, looking 💯

mkaz committed:
Update images to full URLs

chrisvanpatten reviewed (approved):
This is an awesome improvement. Thank you @mkaz!

chrisvanpatten commented on docs/manifest.json:
```diff
@@ -197,6 +197,42 @@
+		"parent": "tutorials"
+	},
+	{
+		"title": "PluginsBackground",
```
@mkaz just noticed the titles in the manifest are incorrect. Did you generate the manifest automatically? Could be an issue in the generator if so, which should get fixed before we merge…

mkaz replied:
I did generate the manifest using `npm run docs:build`

mkaz replied:
Ah, fixed. 

I generated the manifest prior to fixing the header tags and didn't regenerate after. I just ran the script and committed the updated manifest, which has the correct titles. 👍

mkaz committed:
Fix titles

chrisvanpatten commented:
Travis is ¯\_(ツ)_/¯ so I'm going to merge… thanks for working on this @mkaz!
